### PR TITLE
`Development`: Remove code duplication in repositories

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Currently we have a lot of duplicated code in our repositories in the form of `findByIdElseThrow` in nearly every repo. Using a common base interface for all repos and a base class for Proxy generation during runtime this could be removed.
This would reduce boilerplate code and remove variance, as most repos have slightly different variations in the doc, param names and NotNull annotation usage.

### Description

To achieve this I made the following changes:
- Added an `ArtemisJpaRepository` interface, which all other repositories now extend from
- Moved the common `T findByIdElseThrow(ID id)` definition into `ArtemisJpaRepository`
- Added an implementation for `findByIdElseThrow` in the readded `RepositoryImpl` class
- Removed all default `findByIdElseThrow` methods in all repos
- Renamed 2 non-default `findByIdElseThrow` methods to be more accurate and not cause conflicts
- Moved some implementation logic from `DynamicSpecificationRepository` into `RepositoryImpl`
- As a bonus I added a general `getValueElseThrow` method in `ArtemisJpaRepository` with an implementation in `RepositoryImpl`, which takes care of the `.orElseThrow(() -> new EntityNotFoundException(...))` stuff for use in other elseThrow methods. This could then be gradually rolled out or all at once in a follow up.

### Steps for Testing

Prerequisites:
- None

1. Log in to Artemis
2. Click around Artemis a bit

As nothing changed in the behavior and just some logic was moved there are no real changes to test. Thorough code review and passing server/e2e tests are more important.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
- [x] Code Review 3
#### Manual Tests
- [x] Test 1

### Test Coverage

Note: Every new code is either covered a lot of times, or is uncovered, as it is provided for future use (such as `getValueElseThrow` without an id param).
image

### CODE REVIEW

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/base/DynamicSpecificationRepository.java [hunk 1]
```diff
@@ -20,30 +18,45 @@
...
+    /**
+     * Finds an entity by its ID with optional dynamic fetching of associated entities.
+     * Throws an {@link EntityNotFoundException} if the entity with the specified ID does not exist.
+     *
+     * @param id           the ID of the entity to find.
+     * @param fetchOptions a collection of fetch options specifying the entities to fetch dynamically.
+     * @return the entity with the specified ID and the associated entities fetched according to the provided options.
+     */
+    @NotNull
+    @NotNull
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/repository/base/DynamicSpecificationRepository.java]: Guess I got a little over-exited here 😆

File: src/main/java/de/tum/in/www1/artemis/repository/base/RepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+public class RepositoryImpl<T, ID> extends SimpleJpaRepository<T, ID> {
+
+    private final JpaEntityInformation<T, ?> entityInformation;
+
+    /**
+     * Creates a new {@link SimpleJpaRepository} to manage objects of the given {@link JpaEntityInformation}.
+     *
+     * @param entityInformation must not be {@literal null}.
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/repository/base/RepositoryImpl.java]: This was taken from the old implementation. I fixed it.

File: src/main/java/de/tum/in/www1/artemis/repository/base/DynamicSpecificationRepository.java [hunk 3]
```diff
@@ -20,30 +18,45 @@
...
+    default T findByIdWithDynamicFetchElseThrow(ID id, F... fetchOptions) {
+        return findByIdWithDynamicFetchElseThrow(id, Set.of(fetchOptions));
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/repository/base/DynamicSpecificationRepository.java]: That is acutally something I want to investigate. For now I changed it to List to be on the safe side.

**Review by MichaelOwenDyer — DISMISSED** [review]
Good removal of code duplication, thank you for doing this!

**Review by iyannsch — DISMISSED** [review]
Very well thought-though improvement, well done! Did you also think of updating the documentation accordingly? I think this could be a feature which confuses all novice contributors :)

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by krusche — APPROVED** [review]

**Pull request merged**